src/util/any.hpp

```diff
@@ -0,0 +1,230 @@
+        model(const T& other): value(other) {}
+
+        model(T&& other): value(std::move(other)) {}
+
+        interface* copy() override {
+            return new model<T>(*this);
+        }
+
+        const std::type_info& type() override {
+            return typeid(T);
+        }
+
+        void* pointer() override {
+            return &value;
+        }
+
+        const void* pointer() const override {
+            return &value;
+        }
+
+        T value;
+    };
+
+    std::unique_ptr<interface> state_;
+
+protected:
+
+    template <typename T>
+    friend const T* any_cast(const any* operand);
+
+    template <typename T>
+    friend T* any_cast(any* operand);
+
+    template <typename T>
+    T* unsafe_cast() {
+        return static_cast<T*>(state_->pointer());
+    }
+
+    template <typename T>
+    const T* unsafe_cast() const {
+        return static_cast<const T*>(state_->pointer());
+    }
+};
+
+namespace impl {
+
+template <typename T>
+using any_cast_remove_qual = typename
+    std::remove_cv<typename std::remove_reference<T>::type>::type;
+
+} // namespace impl
+
+// If operand is not a null pointer, and the typeid of the requested T matches
+// that of the contents of operand, a pointer to the value contained by operand,
+// otherwise a null pointer.
+template<class T>
+const T* any_cast(const any* operand) {
+    if (operand && operand->type()==typeid(T)) {
+        return operand->unsafe_cast<T>();
+    }
+    return nullptr;
+}
+
+template<class T>
+T* any_cast(any* operand) {
+    if (operand && operand->type()==typeid(T)) {
+        return operand->unsafe_cast<T>();
+    }
+    return nullptr;
+}
+
+template<class T>
+T any_cast(const any& operand) {
+    using U = impl::any_cast_remove_qual<T>;
+    static_assert(std::is_constructible<T, const U&>::value,
+        "any_cast type can't construct copy of contained object");
+
+    auto ptr = any_cast<U>(&operand);
+    if (ptr==nullptr) {
+        throw bad_any_cast();
+    }
+    return static_cast<T>(*ptr);
+}
+
+template<class T>
+T any_cast(any& operand) {
+    using U = impl::any_cast_remove_qual<T>;
+    static_assert(std::is_constructible<T, U&>::value,
+        "any_cast type can't construct copy of contained object");
+
+    auto ptr = any_cast<U>(&operand);
+    if (ptr==nullptr) {
+        throw bad_any_cast();
+    }
+    return static_cast<T>(*ptr);
+}
+
+template<class T>
+T any_cast(any&& operand) {
+    using U = impl::any_cast_remove_qual<T>;
+    static_assert(std::is_constructible<T, U>::value,
+        "any_cast type can't construct copy of contained object");
+
+    auto ptr = any_cast<U>(&operand);
+    if (ptr==nullptr) {
+        throw bad_any_cast();
+    }
+    return static_cast<T>(std::move(*ptr));
+}
+
+// Constructs an any object containing an object of type T, passing the
+// provided arguments to T's constructor.
+template <class T, class... Args>
+any make_any(Args&&... args) {
+    return any(T(std::forward<Args>(args) ...));
+}
```

Comment:
I believe this is not quite the same as what the standard promises, in that this will construct the `T` object and then (generally speaking) use `T`'s move constructor to place it in the wrapped store, while the standard description indicates that the wrapped `T` be constructed in-place.

Solution would be to add comment documenting the difference (or verifying that there is in fact no difference, but I don't think this is true) or implementing the `in_place_t` constructors and invoking them with `make_any` via perfect forwarding.